### PR TITLE
Fill missing sequence IDs up to max sequence ID when recovering from store

### PR DESCRIPTION
**Pull request by s1monw**
Today we might promote a primary and recover from store where after translog
recovery the local checkpoint is still behind the maximum sequence ID seen.
To fill the holes in the sequence ID history this PR adds a utility method
that fills up all missing sequence IDs up to the maximum seen sequence ID
with no-ops.

Relates to #10708
I still work on a test for store recovery to ensure it's called but I think it's ready for review.

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 1]
```diff
@@ -3853,4 +3854,67 @@ private void index(final InternalEngine engine, final int id) throws IOException
...
+        long checkpointOnReplica;
+        try {
+            for (int i = 0; i < docs; i++) {
+                final String docId = Integer.toString(i);
+                final ParsedDocument doc =
+                    testParsedDocument(docId, "test", null, testDocumentWithTextField(), SOURCE, null);
+                Engine.Index primaryResponse = indexForDoc(doc);
+                Engine.IndexResult indexResult = engine.index(primaryResponse);
+                if (randomBoolean()) {
+                    doc.updateSeqID(indexResult.getSeqNo(), 1);
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: why is this needed? doesn't the engine take care of that?

File: core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 2]
```diff
@@ -3853,4 +3854,67 @@ private void index(final InternalEngine engine, final int id) throws IOException
...
+            assertEquals(checkpointOnReplica, replicaEngine.seqNoService().getLocalCheckpoint());
+            recoveringEngine = new InternalEngine(copy(replicaEngine.config(), EngineConfig.OpenMode.OPEN_INDEX_AND_TRANSLOG));
+            assertEquals(numDocsOnReplica, recoveringEngine.getTranslog().totalOperations());
+            recoveringEngine.recoverFromTranslog();
+            assertEquals(maxSeqIDOnReplica, recoveringEngine.seqNoService().getMaxSeqNo());
+            assertEquals(checkpointOnReplica, recoveringEngine.seqNoService().getLocalCheckpoint());
+            assertEquals((maxSeqIDOnReplica+1) - numDocsOnReplica, recoveringEngine.fillSequenceNumberHistory(2));
+            assertEquals(maxSeqIDOnReplica, recoveringEngine.seqNoService().getMaxSeqNo());
+            assertEquals(maxSeqIDOnReplica, recoveringEngine.seqNoService().getLocalCheckpoint());
+            if ((flushed = randomBoolean())) {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: can we snapshot the translog and assert that the noops have the right primary term?

**Review by bleskes — CHANGES_REQUESTED** [review]
Looks great. I one minor request to the test.

File: core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java (hunk 2 above)
Comment (s1monw) [inline core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: ah I had that but remvoed it... good catch...

**Review by bleskes — APPROVED** [review]

File: core/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java [hunk 3]
```diff
@@ -896,6 +898,46 @@ public void testRecoverFromStore() throws IOException {
...
 
+    /* This test just verifies that we fill up local checkpoint up to max seen seqID on primary recovery */
+    public void testRecoverFromStoreWithNoOps() throws IOException {
+        final IndexShard shard = newStartedShard(true);
+        indexDoc(shard, "test", "0");
+        Engine.Index test = indexDoc(shard, "test", "1");
+        // start a replica shard and index the second doc
+        final IndexShard otherShard = newStartedShard(false);
+        test = otherShard.prepareIndexOnReplica(
+            SourceToParse.source(SourceToParse.Origin.PRIMARY, shard.shardId().getIndexName(), test.type(), test.id(), test.source(),
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java]: Origin should REPLICA

File: core/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java [hunk 4]
```diff
@@ -896,6 +898,46 @@ public void testRecoverFromStore() throws IOException {
         closeShards(newShard);
     }
 
+    /* This test just verifies that we fill up local checkpoint up to max seen seqID on primary recovery */
+    public void testRecoverFromStoreWithNoOps() throws IOException {
+        final IndexShard shard = newStartedShard(true);
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java]: I think we can introduce a variant of `indexDoc` called `indexDocOnReplica` which takes a seq# as a parameter. This will remove the need for the extra shard. wdyt?

**Review by bleskes — APPROVED** [review]
Still LGTM. Left a suggestion for the new test.

File: core/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java (hunk 4 above)
Comment (s1monw) [inline core/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java]: I can do that in a sep PR
